### ISSUE TITLE
[A11Y][D9] Provide a "Visually hidden" option for the display of field labels

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5766]
Drupal:
![image](https://user-images.githubusercontent.com/2423362/188629270-5f78aa7f-fecd-47e6-8982-2963ac899f25.png)

Backdrop:
![image](https://user-images.githubusercontent.com/2423362/188629375-3d87e761-b9a8-444e-99e2-345fc933b631.png)

**Comment by klonos** [comment]
Fairly minimal set of changes: https://github.com/backdrop/backdrop/pull/4196

**Comment by klonos** [comment]
...initially I've used `visually_hidden` for the value that's saved in config, the various variables and CSS classes. I have now switched that to `invisible` instead, because it's shorter, and it also matched the `'#title_display' => 'invisible'` we use in our Form API.

**Comment by indigoxela** [comment]
That's actually a great idea, but... as I see it requires to adapt the field.tpl.php.

A handful of contrib modules override it, probably also custom themes or modules. I wonder if we could make this more compatible (probably not).

It's more or less the same situation as with the comment template (recently merged) - which requires all contrib/custom themes that override the template file to update their code, if the theme should support that new feature.

**Comment by klonos** [comment]
>A handful of contrib modules override it, probably also custom themes or modules. I wonder if we could make this more compatible (probably not).

Yeah, I'm not sure how we could do that @indigoxela 🤔 ...I have for some time now been thinking about perhaps adding some sort of hash/versioning in templates and CSS files, so that people can add some logic that would warn them if things changed and their modules/themes needed to be reviewed/revised (because not everyone reads the changelogs). Perhaps something could be shown in the status page too(?) ...but all that is a subject for another, broader issue.

Anyway, another thing that bothers me from a UX perspective is that less technical people seeing the "Hidden" and "Visually hidden" options may start to wonder what the difference between these 2 is. I was contemplating adding some help text at the top/bottom of the table that explains the difference, but I prefer it when things are self-explanatory 🤔 🤔 

Or we could change the user-facing text to something like "Disabled" or "None" (since the current "Hidden" option does not render the label on the page at all), and then use "Hidden" for the newly-introduced option:

- `above` | Above
-  `inline` | Inline
- `hidden` | None
- `invisible` | Hidden

...but having the config value `hidden` be used for something different than what is called "Hidden" in the UI is not ideal at all, and may lead to DX WTFs.

Another thing is that I would like us to encourage/endorse people to be using the new option, for A11Y reasons. So perhaps the user-facing text for the options could be something like this:
- `above` | Above
-  `inline` | Inline
- `invisible` | Visually hidden (recommended)
- `hidden` | Hidden

...or in order to keep things as short as possible, perhaps like this:
- `above` | Above
-  `inline` | Inline
- `invisible` | Visually hidden
- `hidden` | Hidden (discouraged)

Wondering what people think in general UX-wise about these things, so feedback/thoughts welcomed.

**Comment by klonos** [comment]
...anyway, another thing that bothers me from a UX perspective is that less technical people seeing the "Hidden" and "Visually hidden" options may start to wonder what the difference between these 2 is. I was contemplating adding some help text at the top/bottom of the table that explains the difference, but I prefer it when things are self-explanatory 🤔 🤔 

Or we could change the user-facing text to something like "Disabled" or "None" (since the current "Hidden" option does not render the label on the page at all), and then use "Hidden" for the newly-introduced option:

- `above` | Above
-  `inline` | Inline
- `hidden` | None
- `invisible` | Hidden

...but having the config value `hidden` be used for something different than what is called "Hidden" in the UI is not ideal at all, and may lead to DX WTFs.

Another thing is that I would like us to encourage/endorse people to be using the new option, for A11Y reasons. So perhaps the user-facing text for the options could be something like this:
- `above` | Above
-  `inline` | Inline
- `invisible` | Visually hidden (recommended)
- `hidden` | Hidden

...or in order to keep things as short as possible, perhaps like this:
- `above` | Above
-  `inline` | Inline
- `invisible` | Visually hidden
- `hidden` | Hidden (discouraged)

Wondering what people think in general UX-wise about these things, so feedback/thoughts welcomed.

**Comment by klonos** [comment]
...and final though, which I think we should do is to switch the config for the default/teaser displays for the content types that come out of the box to use the new `invisible` option instead of the "legacy" `disabled` one. Does anyone have any objections/concerns if we did that for new installations?

**Comment by indigoxela** [comment]
> ... from a UX perspective is that less technical people seeing the "Hidden" and "Visually hidden" options may start to wonder what the difference between these 2 is

Wow, that's definitely a good point.

"hidden | None" vs invisible | Hidden - another rabbit hole? :rabbit: 

"invisible | Visually hidden (recommended)" hm... this could be read as "better than above or inline" - which is probably not true, at least not always.

"hidden | Hidden (discouraged)" IMO there are valid cases for not adding a field label at all, think of the body field.

**Comment by klonos** [comment]
> ...IMO there are valid cases for not adding a field label at all, think of the body field.

I agree with the rest of the points you are making @indigoxela, but not with this one. That can be served well with `invisible` for sighted people. Non-sighted people need to know that they are viewing the main portion of the page - we shouldn't be assuming that they can tell that or that screen readers automatically differenciate the body field from the rest on the page.

**Comment by indigoxela** [comment]
> the body field may be changed to something like "Biography" or "Review" depending on the content type - it's not necessarily just the vague "Body"

Good point, but wouldn't site builders then display the label (above), anyway, also for sighted people? I'm no expert re accessibility, that's probably already clear.

**Comment by klonos** [comment]
> ...wouldn't site builders then display the label (above), anyway, also for sighted people?

Also good point 🙂 👍🏼 ...also not an accessibility nor UX expert - more empirical knowledge/opinions.

**Comment by stpaultim** [comment]
Just a quick thought:

- above | Above
- inline | Inline
- hidden | Hidden
- invisible | Visually hidden (better for accessibility)

Yeah, I don't like it either. But, it was a thought.

**Comment by olafgrabienski** [comment]
Maybe something like this:

> `invisible` | Visually Hidden (visible for screen readers)

**Comment by indigoxela** [comment]
> invisible | Visually Hidden (visible for screen readers)

Almost... - the "visible" seems a bit odd in that context, though.
Maybe:

- Visually Hidden (exposed to screen readers)
- Visually Hidden (accessible for screen readers)

Is "screen reader" clear enough?

**Comment by olafgrabienski** [comment]
> Visually Hidden (accessible for screen readers)

I like this one! The term "screen reader" is clear enough, in my opinion.

How do we use upper vs. lower case here?

> Visually Hidden vs.
> Visually hidden

**Comment by klonos** [comment]
Thanks for the feedback and great ideas everyone! 🙏🏼 

Throwing another couple of alternatives to the mix for consideration:

- `invisible` | Visually hidden (improved accessibility)
- `invisible` | Shown via assistive technology

A definition from a web search:

>Assistive technology (AT): products, equipment, and systems that enhance learning, working, and daily living for persons with disabilities.

Not married to the above - I think that people that have not been exposed to terms such as "screen reader" and "assistive technology" will be equally unfamiliar with them, and that they would need to Google things.

I still think that long-tern it would be better if we changed the current "Hidden" option to something like "Removed" (as this is what is actually happening) + also move it as the last option in the list (to discourage its usage).  I have also considered "Disabled", but perhaps we should avoid that, as it may contextually clash with the description of the `invisible` option).

> I understand this goal and love it if you come up with an idea. But, sometimes it's ok to add help text and educate the site architect along the way.
> 
> Something like"
> "Visually hidden leaves a hidden label for screen readers."

Yeah, we may end up doing that after all. Anyway, these are just thoughts/suggestions - I'm sure that whatever we end up using, it will be an improvement over the current situation. We may end up exposing the word "Invisible" to the UI after all + adding some help text:

`above` | Above
`inline` | Inline
`invisible` | Invisible*
`hidden` | Removed

*Hidden, but still visible via screen readers and other assistive technology.

> How do we use upper vs. lower case here?

We capitalize things like module/theme names etc. - not user-facing text and options. It should be "Visually hidden".

**Comment by indigoxela** [comment]
> ... Shown via assistive technology

I love the "assistive technology" part, but find the "shown" a bit odd, as to me show is connected to visible (not a native speaker, though, so my impression may be wrong).

> I still think that long-tern it would be better if we changed the current "Hidden" option to something like "Removed" (as this is what is actually happening)

Agreed, but possibly in a follow-up, as the "hidden" setting is used in more contexts than just field labels, and we should try to stay consistent. And even if we only change it for field labels, we'd have to carefully consider possible impacts on existing code outside core. But that shouldn't block the enhancement, this issue provides.

**Comment by klonos** [comment]
> ...but find the "shown" a bit odd

Yeah, you're right about that 👍🏼
